File: api_gateway/services/tests/fixtures/sample.ts
/**
 * Sample TypeScript module for testing the code parser.
 *
 * This module contains various TypeScript constructs to test the parser's
 * ability to extract interfaces, types, classes, functions, and variables.
 */

import { EventEmitter } from "events";

/**
 * Configuration options for services.
 */
export interface ServiceConfig {
  /** Service name identifier */
  name: string;
  /** Base URL for API calls */
  baseUrl: string;
  /** Request timeout in milliseconds */
  timeout?: number;
  /** Enable debug logging */
  debug?: boolean;
}

/**
 * Response structure from API calls.
 */
export interface ApiResponse<T = unknown> {
  /** Response data */
  data: T;
  /** HTTP status code */
  status: number;
  /** Response headers */
  headers: Record<string, string>;
}

/**
 * Service status type.
 */
export type ServiceStatus = "idle" | "running" | "stopped" | "error";

/**
 * Union type for identifiers.
 */
export type Identifier = string | number;

/**
 * Callback function type for event handlers.
 */
export type EventCallback<T = void> = (event: T) => void;

/** Default timeout value in milliseconds */
export const DEFAULT_TIMEOUT = 30000;

/** Maximum retry attempts */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const MAX_RETRIES = 3;

/**
 * Simple utility function to add numbers.
 *
 * @param a - First number
 * @param b - Second number
 * @returns Sum of a and b
 */
export function add(a: number, b: number): number {
  return a + b;
}

/**
 * Async function to fetch data from URL.
 *
 * @param url - The URL to fetch
 * @param options - Optional fetch options
 * @returns Promise resolving to API response
 */
export async function fetchData<T>(
  url: string,
  options?: RequestInit
): Promise<ApiResponse<T>> {
  const response = await fetch(url, options);
  const data = await response.json();
  return {
    data: data as T,
    status: response.status,
    headers: Object.fromEntries(response.headers.entries()),
  };
}

/**
 * Arrow function assigned to variable.
 */
export const multiply = (a: number, b: number): number => a * b;

/**
 * Async arrow function.
 */
export const delay = async (ms: number): Promise<void> => {
  return new Promise((resolve) => setTimeout(resolve, ms));
};

/**
 * Abstract base class for services.
 */
export abstract class BaseService {
  /** Service name */
  protected readonly name: string;

  /** Service status */
  protected status: ServiceStatus = "idle";

  /**
   * Create a new service instance.
   *
   * @param name - Service name
   */
  constructor(name: string) {
    this.name = name;
  }

  /**
   * Get the service name.
   *
   * @returns Service name string
   */
  getName(): string {
    return this.name;
  }

  /**
   * Get the current status.
   *
   * @returns Current service status
   */
  getStatus(): ServiceStatus {
    return this.status;
  }

  /**
   * Abstract method to start the service.
   */
  abstract start(): Promise<void>;

  /**
   * Abstract method to stop the service.
   */
  abstract stop(): Promise<void>;
}

/**
 * Concrete service implementation with configuration.
 *
 * @example
 * ```typescript
 * const service = new ConfiguredService({
 *   name: "api",
 *   baseUrl: "https://api.example.com",
 *   timeout: 5000,
 * });
 * await service.start();
 * ```
 */
export class ConfiguredService extends BaseService {
  /** Service configuration */
  private config: ServiceConfig;

  /** Event emitter for service events */
  private emitter: EventEmitter;

  /**
   * Create a configured service.
   *
   * @param config - Service configuration options
   */
  constructor(config: ServiceConfig) {
    super(config.name);
    this.config = config;
    this.emitter = new EventEmitter();
  }

  /**
   * Get the service configuration.
   *
   * @returns Current configuration object
   */
  getConfig(): ServiceConfig {
    return { ...this.config };
  }

  /**
   * Update configuration options.
   *
   * @param updates - Partial configuration updates
   */
  updateConfig(updates: Partial<ServiceConfig>): void {
    this.config = { ...this.config, ...updates };
    this.emitter.emit("configChanged", this.config);
  }

  /**
   * Start the service.
   */
  async start(): Promise<void> {
    this.status = "running";
    this.emitter.emit("started");
  }

  /**
   * Stop the service.
   */
  async stop(): Promise<void> {
    this.status = "stopped";
    this.emitter.emit("stopped");
  }

  /**
   * Subscribe to service events.
   *
   * @param event - Event name
   * @param callback - Event handler callback
   */
  on<T>(event: string, callback: EventCallback<T>): void {
    this.emitter.on(event, callback);
  }

  /**
   * Static factory method to create service from config object.
   *
   * @param config - Configuration object
   * @returns New ConfiguredService instance
   */
  static fromConfig(config: ServiceConfig): ConfiguredService {
    return new ConfiguredService(config);
  }
}

/**
 * Enum for log levels.
 */
export enum LogLevel {
  DEBUG = 0,
  INFO = 1,
  WARN = 2,
  ERROR = 3,
}

/**
 * Simple logging utility class.
 */
export class Logger {
  private level: LogLevel;
  private prefix: string;

  constructor(prefix: string = "", level: LogLevel = LogLevel.INFO) {
    this.prefix = prefix;
    this.level = level;
  }

  /**
   * Log a debug message.
   */
  debug(message: string): void {
    if (this.level <= LogLevel.DEBUG) {
      console.debug(`${this.prefix}[DEBUG] ${message}`);
    }
  }

  /**
   * Log an info message.
   */
  info(message: string): void {
    if (this.level <= LogLevel.INFO) {
      console.info(`${this.prefix}[INFO] ${message}`);
    }
  }

  /**
   * Log a warning message.
   */
  warn(message: string): void {
    if (this.level <= LogLevel.WARN) {
      console.warn(`${this.prefix}[WARN] ${message}`);
    }
  }

  /**
   * Log an error message.
   */
  error(message: string): void {
    console.error(`${this.prefix}[ERROR] ${message}`);
  }
}

File: api_gateway/services/tests/fixtures/sample.py
"""
Sample Python module for testing the code parser.

This module contains various Python constructs to test the parser's
ability to extract functions, classes, variables, and other entities.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict, Optional, TypeVar

# Module-level constant
DEFAULT_TIMEOUT: int = 30

# Module-level variable without annotation
MAX_RETRIES = 3

# Type variable for generics
T = TypeVar("T")


def simple_function(x: int, y: int) -> int:
    """
    Add two numbers together.

    Args:
        x: First number
        y: Second number

    Returns:
        Sum of x and y
    """
    return x + y


def function_with_defaults(
    name: str,
    count: int = 10,
    enabled: bool = True,
    options: Optional[Dict[str, Any]] = None,
) -> str:
    """Function with default parameter values."""
    return f"{name}: {count}, enabled={enabled}"


async def async_function(url: str) -> Dict[str, Any]:
    """
    Asynchronous function that fetches data.

    Args:
        url: The URL to fetch

    Returns:
        Response data as dictionary
    """
    return {"url": url, "status": "ok"}


@dataclass
class SimpleDataclass:
    """A simple dataclass with typed fields."""

    name: str
    value: int
    enabled: bool = True


class BaseService:
    """Base class for services."""

    def __init__(self, name: str) -> None:
        """Initialize the service with a name."""
        self.name = name

    def get_name(self) -> str:
        """Return the service name."""
        return self.name


class AdvancedService(BaseService):
    """
    Advanced service with more features.

    This class extends BaseService with additional functionality
    including configuration management and async operations.
    """

    def __init__(
        self,
        name: str,
        config: Optional[Dict[str, Any]] = None,
        timeout: int = DEFAULT_TIMEOUT,
    ) -> None:
        """
        Initialize the advanced service.

        Args:
            name: Service name
            config: Optional configuration dictionary
            timeout: Request timeout in seconds
        """
        super().__init__(name)
        self.config = config or {}
        self.timeout = timeout

    @property
    def is_configured(self) -> bool:
        """Check if the service has configuration."""
        return bool(self.config)

    @staticmethod
    def validate_config(config: Dict[str, Any]) -> bool:
        """
        Validate a configuration dictionary.

        Args:
            config: Configuration to validate

        Returns:
            True if valid, False otherwise
        """
        return "name" in config

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "AdvancedService":
        """
        Create an instance from a dictionary.

        Args:
            data: Dictionary with service data

        Returns:
            New AdvancedService instance
        """
        return cls(
            name=data.get("name", "default"),
            config=data.get("config"),
            timeout=data.get("timeout", DEFAULT_TIMEOUT),
        )

    async def fetch_data(self, endpoint: str) -> Dict[str, Any]:
        """
        Fetch data from an endpoint asynchronously.

        Args:
            endpoint: API endpoint path

        Returns:
            Response data
        """
        return {"endpoint": endpoint, "service": self.name}

    class NestedHelper:
        """A nested helper class."""

        def __init__(self, parent: "AdvancedService") -> None:
            self.parent = parent

        def get_parent_name(self) -> str:
            """Get the parent service name."""
            return self.parent.name


# Annotated variable with complex type
service_registry: Dict[str, BaseService] = {}


def _private_helper(data: Any) -> str:
    """Private helper function."""
    return str(data)


def function_with_varargs(
    *args: int,
    multiplier: int = 1,
    **kwargs: str,
) -> Dict[str, Any]:
    """
    Function with *args, keyword-only args, and **kwargs.

    Args:
        *args: Variable positional arguments
        multiplier: Keyword-only argument with default
        **kwargs: Variable keyword arguments

    Returns:
        Dictionary with processed arguments
    """
    return {
        "sum": sum(args) * multiplier,
        "kwargs": kwargs,
    }


def outer_function(x: int) -> int:
    """
    Outer function that contains a nested function.

    Args:
        x: Input value

    Returns:
        Processed value from nested function
    """
    def inner_function(y: int) -> int:
        """Nested inner function that doubles the value."""
        return y * 2

    return inner_function(x) + x
